src/jdk.httpserver/share/classes/com/sun/net/httpserver/Filter.java

```diff
@@ -140,4 +142,117 @@ public abstract void doFilter (HttpExchange exchange, Chain chain)
+     * filter in the chain).
+     *
+     * @apiNote
+     * A beforeHandler filter is typically used to examine or modify the
+     * exchange state before it is handled. The filter operation is executed
+     * before {@link Filter.Chain#doFilter(HttpExchange)} is invoked, so before
+     * any subsequent filters in the chain and the exchange handler are
+     * executed. The response is commonly sent by the exchange handler.
+     * The filter is only expected to send the response in the uncommon case
+     * that the exchange will not be handled after the filter is executed.
+     *
+     * <p> Example of adding a response header to all responses:
+     * <pre>{@code
+     *     var filter = Filter.beforeHandler("Add response header X-Content-Type-Options",
+     *                 (var e) -> e.getResponseHeaders().set("X-Content-Type-Options", "nosniff"));
+     *     var server = HttpServer.create(new InetSocketAddress(0), 10);
+     *     server.createContext("/", new SomeHandler()).getFilters().add(filter);
+     *     server.start();
+     * }</pre>
+     *
+     * @param description the description of the returned filter
+     * @param operation the operation of the returned filter
+     * @return a filter
+     * @throws NullPointerException if any argument is null
+     * @since 17
+     */
+    public static Filter beforeHandler(String description,
+                                       Consumer<HttpExchange> operation) {
+        Objects.requireNonNull(description);
+        Objects.requireNonNull(operation);
+        return new Filter() {
+            @Override
+            public void doFilter(HttpExchange exchange, Chain chain) throws IOException {
+                operation.accept(exchange);
+                chain.doFilter(exchange);
+            }
+            @Override
+            public String description() {
+                return description;
+            }
+        };
+    }
+
+    /**
+     * Returns a post-processing {@code Filter} with the given description and
+     * operation.
+     *
+     * <p>The {@code description} string describes the returned filter. The
+     * {@link Consumer operation} is the effective implementation of the filter.
+     * It is executed for each {@code HttpExchange} after invoking either the
+     * next filter in the chain or the exchange handler (if this filter is the
+     * final filter in the chain).
+     *
+     * @apiNote
+     * An afterHandler filter is typically used to examine the exchange state
+     * rather than modifying it. The filter operation is executed after
+     * {@link Filter.Chain#doFilter(HttpExchange)} is invoked, this means any
+     * subsequent filters in the chain and the exchange handler have been
+     * executed and the response has commonly been sent. The filter is only
+     * expected to send the response in the uncommon case that the exchange
+     * has not been handled before the filter is executed.
```

Comment:
Maybe we could rephrase it here too:

```
 * executed and the response has commonly been sent. The filter 
 * {@code operation} is typically not expected to handle the exchange or 
 * {@linkplain HttpExchange#sendResponseHeaders(int, long) send the response headers}
 * as doing so is likely to fail, since the exchange is expected to have already been handled 
 * at this point.
```

I would like to make it more obvious that `afterHandler` and `beforeHandler` are not the right API to handle the exchange. I have the feeling that saying that "it is only expected to send ... " does not convey that strongly enough.